### ISSUE TITLE
Suggestion to improve adminbundle SCSS

### ISSUE THREAD

**Issue by Arjen-Smit** [sulu/sulu#2926]
| Q | A |
| --- | --- |
| Bug? | no |
| New Feature? | no |
| Sulu Version | any |
| Browser Version | any |

Currently the best way to add custom branding to the sulu adminbundle is by loading a stylesheet with additional overwriting rules after loading the default stylesheet. Although this works in most cases I think there is a more elegant way to approach this.

There is a way to make some changes that wont break any current styling but that can be used to easily overwrite colors or images.

If you take for example this small snippet from base.scss

``` scss
.initial-loader {
    $dimension: 100px;
    width: $dimension;
    height: $dimension;
    @include centerBlock();
    .bouncer {
        background: $grayLighter;
    }
}
```

Could be changed to:

``` scss
// top of the file
$inital-loader-background: $greyLighter !default;


.initial-loader {
    $dimension: 100px;
    width: $dimension;
    height: $dimension;
    @include centerBlock();
    .bouncer {
        background: $inital-loader-background;
    }
}
```

The !default will use the value $greyLighter unless the variable is already set before this line.
This way you can add a file to your own project that looks something like this:

``` scss
$inital-loader-background: #b6b7ca;

@import '/path-to-sulu-bundle/scss/main';

```

This will result in a css file that is simular to the original adminbundle css file but with other colors (or images). Another great thing about this way of working is that a structural change in the scss of the admin bundle won't cause any conflicts with the overwrites that are normally done afterwards since they simply use the new structure but with correct already defined variables.

Any comments or suggestion to this sugestion are welcome, and if we can agree on this I don't mind doing some of the refactoring to make all colors and image configurable.

**Comment by danrot** [comment]
Really like the idea, using colors directly is not the best idea anyway. But I would use the camel case notation instead of connection the words using dashes, since it's already like that for the color itself (`greyLighter`).

**Comment by Arjen-Smit** [comment]
I personally prefer hyphenated variables but I'm fine with camelCase as well. I'll start working on a PR but if anyone has any more suggestions feel free to discuss them.

**Comment by danrot** [comment]
For me consistency is more important than personal taste 😃

**Comment by dvdweel** [comment]
@danrot 

Maybe it is a good idea to think about the variable names. 

If we take for example this snippet from login.scss

```css
.success-overlay {
            height: 100%;
            width: 100%;
            position: relative;
            z-index: -10;

            background-color: $green; 
        }
```

And would change it the way @Arjen-Smit  is suggesting. You can end up with very long variable names.

```css
$suluLoginContentContainerSuccessOverlayBackgroundColor: $green !default;

.success-overlay {
            height: 100%;
            width: 100%;
            position: relative;
            z-index: -10;

            background-color: $suluLoginContentContainerSuccessOverlayBackgroundColor;
        }
```


Maybe it is better to shorten the names so it would be easier to understand. There are some parts of the variable that are unnecessary and without them we still know what value we want to change.

Personally I would change
```css
$suluLoginContentContainerSuccessOverlayBackgroundColor: $green !default;
```

to

```css
$LoginSuccessOverlayBackgroundColor: $green !default;
```

This way we still can overwrite the values without using very long variable names.
What do you think will be the best approach?

**Comment by danrot** [comment]
From what I see in @Arjen-Smit suggestion the variable name doesn't get so long. I would also prefer your shorter version, but it should start lower case: `$loginSuccessOverlayBackgroundColor` should do the job.

**Comment by Arjen-Smit** [comment]
that would work for me

**Comment by dvdweel** [comment]
@Arjen-Smit and I have been working on improving the SCSS and changed the variables so it is possible to overwrite them. There will be a PR soon.

In order to change for example the sulu logo it is necessary to make variables in husky overwritable as well. 
We would like to use the same tactic here as we did for the changes in sulu (see snippets on opening post).
Should we open an issue for that on the husky repository or can it be discussed in here ?

**Comment by danrot** [comment]
I would say that can also be discussed here. Is it only about the Sulu Logo?

**Comment by Arjen-Smit** [comment]
In my opinion we should consider the option to actually make everything in husky configureable. That way we can generate completely custom themes for sulu, and also make husky better implementable into other projects that might not like the colours that are now hardcoded into it. 

I was thinking about opening another issue when this issue was done to make the admin stylesheet configureable. one of the advantages of that is we don't need to overwrite the default admin bundle and another advantage is that we can for use different stylesheets for different environments. This can make is easier for the user to see what environment he is using.

**Comment by danrot** [comment]
Yeah, I would also say we start with the Sulu Logo. What exactly do you mean admin stylesheet you want to work on after this issue is closed?

And although I agree with most of what you are saying I can't imagine that we can use different colors to match different environments, because we also have the compiled scss in version control. So we can't really use the Symfony environment to change that scss variable...

**Comment by Arjen-Smit** [comment]
Sorry for the late reply, but I had the following solution in mind.
By default the admin bundle includes both husky.min.css and main.css, these files combined make up the normal styling of the admin panels. In my opinion they could be made configureable in the /app/config/admin/config.yml replacing both files by a single file with all required custom styling (but only if configured). If that is possible it's just as easy to configure a different css file in /app/config/admin/config_prod.yml making the styling environment aware. But since that is not directly connected to this issue I wanted to create a separate issue for that. 

Below is the simplest example that can already be implemented if the current PR in husky is excepted.
```
// admin.scss

$logoBackgroundImage: '/assets/images/logo.png'
@include huskyfiles;
@include sulusfiles 
```

```
// admin-dev.scss

$logoBackgroundImage: '/assets/images/logo-dev.png'
@include huskyfiles;
@include sulusfiles 
```

**Comment by danrot** [comment]
Well, the problem is that we have a separate grunt task, which compiles the scss to css. And this grunt task is not aware of the symfony environment at all. This means we would have to create two separate css files with the grunt task, with two different outputs, save both outputs and configure them in the configuration files. Sounds like a lot of hassle to me... Wouldn't it be enough to show a red label on top of the application, when you open it in production mode? That's something that could probably be easily implemented in a separate bundle, if you really want to have something like this.

**Comment by Arjen-Smit** [comment]
I think this conversation might be focussed a bit to much on the example's given instead of the functionality purposed. I'm purposing that we can overwrite the default files that are currently hardcoded in a twig template. This way we don't have to change anything else in Sulu and still be able to load a custom stylesheet. The multi environment configuration is not a part of that proposal, but more something that could be achieved by the same change.

**Comment by danrot** [comment]
Ok, but you are still suggesting a bigger change in the infrastructure here, right?

Your suggestion would conclude that we are not using the built css files of husky in Sulu anymore, but also have to copy over the scss files, right? Because otherwise we can't change the values of the scss variables in Sulu anymore, or am I overlooking something?

**Comment by danrot** [comment]
@Arjen-Smit Is that issue solved for you now? We merged all the PRs from @dvdweel, and we are not using colors directly anymore.

**Comment by danrot** [comment]
This has been fixed for Sulu 1.6, and in Sulu 2.0 we also used variable names, so it should be possible to override them as well.
